src/tree.rs

```diff
@@ -217,41 +520,267 @@ impl PartialEq for Tree {
+            },
+            (EntryParents::Many(old_parents), EntryParents::Many(mut new_parents)) => {
+                new_parents.extend(old_parents);
+                Some(EntryParents::Many(new_parents))
+            },
+            _ => None,
+        }
+    }
+
+    fn indices(&self) -> Vec<Index> {
+        match self {
+            EntryParents::Root => Vec::new(),
+            EntryParents::One(parent_from) => match parent_from {
+                EntryParentFrom::Children(parent_index) => vec![*parent_index],
+                EntryParentFrom::Item(_) => Vec::new(),
+            },
+            EntryParents::Many(parents_from) => {
+                parents_from.iter().filter_map(|parent_from| match parent_from {
+                    EntryParentFrom::Children(index) => Some(*index),
+                    EntryParentFrom::Item(_) => None,
+                }).collect()
+            }
+        }
+    }
+
+    fn guids(&self) -> Vec<Guid> {
+        match self {
+            EntryParents::Root => Vec::new(),
+            EntryParents::One(parent_from) => match parent_from {
+                EntryParentFrom::Children(_) => Vec::new(),
+                EntryParentFrom::Item(guid) => vec![guid.to_owned()],
+            },
+            EntryParents::Many(parents_from) => {
+                parents_from.iter().filter_map(|parent_from| match parent_from {
+                    EntryParentFrom::Children(_) => None,
+                    EntryParentFrom::Item(guid) => Some(guid.to_owned()),
+                }).collect()
+            }
+        }
+    }
+}
+
+
+/// Describes where an entry's parent comes from.
+///
+/// `EntryParentFrom` notes inconsistencies like orphans, multiple parents,
+/// and parent-child disagreements, which the `Merger` uses to resolve invalid
+/// structure and produce a consistent tree.
+#[derive(Clone, Debug)]
+enum EntryParentFrom {
+    /// The entry's parent references the entry in its `children`, meaning we
+    /// know the index of the parent.
+    Children(Index),
+
+    /// The entry's parent comes from its `parentid`. We can only store the
+    /// GUID and not the index because the tree might not have an entry for the
+    /// `parentid` yet.
+    Item(Guid),
+}
+
+#[derive(Clone, Copy, Debug)]
+enum Divergence {
+    /// The node's structure is already correct, and doesn't need to be
+    /// reuploaded.
+    Ok,
+
+    /// The node exists in multiple parents, or is a reparented orphan.
+    /// The merger should reupload the node.
+    Diverged,
 }
 
 /// A convenience wrapper around `Entry` that dereferences to the containing
 /// item, and follows indices for parents and children.
 #[derive(Clone, Copy, Debug)]
-pub struct Node<'t>(&'t Tree, &'t Entry);
+pub struct Node<'t>(&'t Tree, &'t Entry, Divergence);
 
 impl<'t> Node<'t> {
+    /// Returns an iterator for all resolved children of this node, including
+    /// reparented orphans.
     pub fn children<'n>(&'n self) -> impl Iterator<Item = Node<'t>> + 'n {
```

Comment:
Since I move `self` into the closure, and handle orphans and `child_indices` slightly differently...I guess I could zip the orphans and `child_indices` with an `iter::repeat` that passes along a tag (`Option<Divergence>`?).

_Or_, since everything in `orphan_indices_by_parent_guid` should _already_ be `Diverged`, I can probably just `chain` orphans before `filter_map`.